chai/source.py

```diff
@@ -0,0 +1,169 @@
+    """Find the dependencies from a line in a module"""
+    # Drop the prefix parts of the lines
+    rest = re.sub(EXTENDS_LINE_PREFIX_RE, "", re.sub(INSTANCE_LINE_PREFIX_RE, "", line))
+    return [
+        non_empty for non_empty in (dep.strip() for dep in rest.split(",")) if non_empty
+    ]
+
+
+def _get_dep_from_instance_line(line: str) -> List[str]:
+    """Find the dependencies from an INSTANCE declaration"""
+    rest = re.sub(INSTANCE_LINE_PREFIX_RE, "", line)
+    # The dependency will be the first word in the line...
+    dep = next(
+        (d for d in rest.split(" ") if d),
+        None,  # or, None if we don't have any words
+    )
+    if dep is None:
+        return []
+    else:
+        return [dep]
+
+
+def _get_module_deps(module: str) -> List[str]:
+    """Extract the module dependencies from a TLA module
+
+    For the grammar of TLA module imports, see
+
+    - EXTENDS: https://github.com/tlaplus-community/tlaplus-standard/blob/3cbb0c251d63d5365f71aa0b3fea0719b6879edd/grammar/TLAPlus2Grammar.tla#L83 # noqa
+    - INSTANCE: https://github.com/tlaplus-community/tlaplus-standard/blob/3cbb0c251d63d5365f71aa0b3fea0719b6879edd/grammar/TLAPlus2Grammar.tla#L141-L144 # noqa
+
+    And for examples of both: http://lamport.azurewebsites.net/tla/newmodule.html#Section2
+    """
+    deps = []
+    # Comma separated deps may extend over many lines
+    in_comma_sep_deps = False
+    for line in module.splitlines():
+        if not in_comma_sep_deps and re.search(EXTENDS_LINE_PREFIX_RE, line):
+            in_comma_sep_deps = True
+
+        if not in_comma_sep_deps and re.search(INSTANCE_LINE_PREFIX_RE, line):
+            if "," in line:
+                in_comma_sep_deps = True
+            else:
+                new_deps = _get_dep_from_instance_line(line)
+                deps.extend(new_deps)
+
+        if in_comma_sep_deps:
+            new_deps = _get_comma_separated_deps(line)
+            # Check if we have a blank line,
+            # in which case we don't add anything and keep searching
+            if new_deps:
+                deps.extend(new_deps)
+                # If the line ends in a comma, we will have more deps to come
+                if not line.rstrip().endswith(","):
+                    in_comma_sep_deps = False
+
+    return deps
+
+
+def _load_deps_of_tla_file(tla_module: Path) -> List[str]:
+    """Load all the found dependencies on disk for TLA+ module
+
+    Return:
+        a list of the contents of the dependenceis
+    """
+    print(">>>", tla_module)
+    content = tla_module.read_text()
+    deps = _get_module_deps(content)
+    print(deps)
+    return [
+        dep
+        for f in tla_module.parent.resolve().iterdir()
+        if f.stem in deps and f.suffix.lstrip(".") == TLA_SUFFIX
+        for dep in (_load_deps_of_tla_file(f) + [f.read_text()])
+    ]
+
+
+class Source:
+    """
+    A source from which the client can load data
+    """
```

Comment:
It took me some time to understand the interface of this class.
afaict this supports two different kinds of callers?

1. The caller knows what kind of `Input` it's loading. These use `Source/__init__` and `of_file[_load_deps]`.
2. The caller doesn't and has to go through `load_input`, then call `Source/__init__` on the results.

It seems that we could have a more uniform interface by providing a `@classmethod` that dynamically dispatches based on the `Input` type like `load_input`, but actually returns a `Source`?

Alternatively, I'd suggest to expand the docstring:

```suggestion
class Source:
    """
    In-memory representation of a TLA+ module and it's dependencies as strings.

    To initialize from filesystem objects, use the class methods `of_file`
    and `of_file_load_deps`.
    If the kind of `Input` is unknown, use `load_input` to transparently
    convert an `Input` into a string.
    """
```

Reply:
Yet another thought: Should `Input` be a class that has a `load() -> str` method?

Reply:
Good questions here.
 
> It seems that we could have a more uniform interface by providing a @classmethod that dynamically dispatches based on the Input type like load_input, but actually returns a Source?

Two things have motivated moving away from an overloaded constructor working on both `str`s and `Path`s:

- Andrey voiced a preference against that approach in our sync with MBT a few weeks back 
- `str` inputs *must* be accompanied by a format specification (because IMO it would be too much magic to try to infer the format by inspecting the string), but this is unnecessary and even introduces the potential for errors to support formats for files, for which the format is derived from the extension.

So we really need two separate constructors to safely construct the source.

The `load_input` really doesn't have anything to do with the `Source` class now, it's just a convenience function, and maybe should be made private... I'll think it over a bit. 

I'll incorporate something like your docstring update. Thanks!

Reply:
Sorry for the messy work here: it turns out `load_input` was really only "needed" due to the earlier design decisions made for the `trans_explorer`, prior to work started on the `cmd_executor`. The subsequent work has changed the RPC approach quite a bit, so we should followup with reworking the `trans_explorer` code. I've opened an issue to track this:  #33

Reply:
See 

- [c88eae8](https://github.com/informalsystems/apalache-chai/pull/32/commits/c88eae8e7eea8c4d219489f63d764042986464e1) for the "deprecation" of the `load_input` method 
- 85fa4ccf5807e5155acfaa4280abda4f5bbd26a7 for inclusion of your suggested docstring

Reply:
Looks good!
I think I was really just confused by the 3 class methods and how to combine them to construct a source; thanks for addressing this!

Reply:
You'd definitely caught a messy bit needing cleaning up. Thanks!